src/Autofac/Core/Resolving/InstanceLookup.cs

```diff
@@ -81,19 +83,43 @@ public object Execute()
             _executed = true;
 
             object? decoratorTarget = null;
-            object instance = ComponentRegistration.Sharing == InstanceSharing.None
-                ? Activate(Parameters, out decoratorTarget)
-                : _activationScope.GetOrCreateAndShare(ComponentRegistration.Id, () => Activate(Parameters, out decoratorTarget));
+
```

Comment:
I'll need someone else to take a look at this. I haven't spent much time in the decorator stuff and could miss something important. Ideally @alexmg but @alistairjevans or @alsami may have some insight as well?

Reply:
@tillig I haven't worked with that code so far but I could dig into it if no one has time that also already has a better understanding of it.

Reply:
I'd welcome any available help on this. I know I won't have time to dive deeper.

Reply:
I'll take a look at this one.

Reply:
The fix looks good and does require the `GetOrCreateAndShare` method to be split up. I don't think there is any point leaving the old method behind and unused either. There is no mechanism in place for you to swap out the `ISharingLifetimeScope` implementation and internally we create instances of `LifetimeScope` and pass them through internal classes. I don't see a reason why anyone would be implementing this interface given the constraints.

This is definitely a breaking change but I don't think it is one that will have much of an impact due to it not being an extensibility point, but these things can certainly play out in unexpected ways. I would be fine with trying to get away with a minor release for this one.

A few suggestions:

- Remove the `GetOrCreateAndShare` method from `ISharingLifetimeScope` and implementation in `LifetimeScope` (not used and we are breaking things anyway)
- Rename `Create` method to `CreateSharedInstance`
- Rename `Get` to `TryGetSharedInstance` with an out parameter and return a `bool` instead of `null` when not found

It would be good to consider how we can actually make these methods internal and remove them from the public API but that's probably a job for a separate issue.

Reply:
Thank for your suggestions @alexmg I have updated the code. :)